opass: allow listing subset of paths

gtback opened the pull request:
Add an argument to the `opass list` command. By default it is `None`,
which retains the current behavior of listing all opass secrets. But if
specified, `opass list` will only list paths that begin with that
substring.

gtback committed:
opass: allow listing subset of paths

Add an argument to the `opass list` command. By default it is `None`,
which retains the current behavior of listing all opass secrets. But if
specified, `opass list` will only list paths that begin with that
substring.

kfogel commented on opass:
```diff
@@ -300,7 +300,7 @@ def main():
         sys.exit(1)
 
-    def list_paths():
+    def list_paths(subpath=None):
```
When adding an argument to a method, please also add documentation for that argument.  (The other functions and methods in `opass` show examples of how.)

gtback replied:
Fixed in d15ee67e7aa7dea970c29ebf0c20e5bc81b5cd8a

kfogel commented on opass:
```diff
@@ -315,12 +315,17 @@ def main():
+                if filename.endswith(".gpg"):
+                    # Remove leading slash and trailing ".gpg"
+                    path = '{}/{}'.format(dirname, filename)[1:-4]
+                    if subpath and not path.startswith(subpath):
```
This conditionalizes on *any* prefix, not necessarily a path-component-wise prefix.  In other words, if the user were to pass "far" as the argument, then the path "farmer/brown" would be considered a match.  Is this the desired behavior?

(Documenting the behavior of the new `subpath` parameter would have forced this question into consideration from the start, because to document it we'd have to decide what the precise behavior should be :-). )

gtback replied:
Yep, I considered this but didn't document it. I decided that a strict string prefix match was easier to implement than one that considered path components as atomic units. I'll document this, but if you think the other behavior is preferable and worth the additional effort (I don't think either `fnmatch` nor `glob` does exactly what we need), I can work on that.

gtback replied:
I'm actually going to change the argument name to `prefix` to make that more clear.

gtback replied:
Renamed in 9793b32b5422cb673d4633844160c2dd085abbbf and documented in d15ee67e7aa7dea970c29ebf0c20e5bc81b5cd8a

kfogel replied:
I think it's worth the additional effort -- it's really a correctness issue: the intended/desired behavior is specifically path-component-wise not string-wise.  Also, I think the `os.path` library in Python has stuff to help with this?  See `os.path.commonprefix()` and other stuff in [os.path](https://docs.python.org/3/library/os.path.html).

gtback replied:
Ah, yep, the `os.path` module should work.

kfogel replied:
Great!  Yeah, that module is full of useful tricks.

kfogel replied:
Just double-checking that ball is still in your court, @gtback (for converting string-prefix behavior to pathwise-prefix behavior).  No rush; I just wanted to make sure I had a correct understanding of the current status of this PR.

kfogel reviewed (changes requested):
Following up on my in-diff comments:

The `__doc__` string at the top of the file also needs to be updated to describe the new behavior.

gtback committed:
opass: Rename "subpath" to "prefix"

This more accurately describes how the matching is done, using literal
string-prefix semantics rather than considering path semantics.

gtback committed:
opass: Document PREFIX argument to 'list' command
